# 1.0 Data Cleaning Walkthrough

## 1.1 Introduction

At many points in your career, you'll need to be able to build complete, end-to-end data science projects on your own. **Data science projects usually consist of one of two things:**

- An exploration and analysis of a set of data. One example might involve analyzing donors to political campaigns, creating a plot, and then sharing an analysis of the plot with others.
- An operational system that generates predictions based on data that updates continually. An algorithm that pulls in daily stock ticker data and predicts which stock prices will rise and fall would be one example.

You'll find the ability to create data science projects useful in several different contexts:

- Projects will help you build a portfolio, which is critical to finding a job as a data analyst or scientist.
- Working on projects will help you learn new skills and reinforce existing concepts.
- Most "real-world" data science and analysis work consists of developing internal projects.
- Projects allow you to investigate interesting phenomena and satisfy your curiosity.

Whether you aim to become a data scientist or analyst or you're just curious about the world, building projects can be immensely rewarding.

**In this section, we'll walk through the first part of a complete data science project, including how to acquire the raw data.** The project will focus on exploring and analyzing a data set. We'll develop our **data cleaning** and **storytelling skills**, which will enable us to build complete projects on our own.

We'll focus **primarily on data exploration** in this section. We'll also combine several messy data sets into a single clean one to make analysis easier. Over the next few missions, we'll work through the rest of our project and perform the actual analysis.

**The first step in creating a project is to decide on a topic**. You want the topic to be something you're interested in and motivated to explore. It's very obvious when people are making projects just to make them, rather than out of a genuine interest in the topic.

Here are two ways to go about finding a good topic:

- Think about what sectors or angles you're really interested in, then find data sets relating to those sectors.
- Review several data sets, and find one that seems interesting enough to explore.

Whichever approach you take, you can start your search at these sites:

- [Data.gov](https://www.data.gov/) - A directory of government data downloads
- [/r/datasets](https://reddit.com/r/datasets) - A subreddit that has hundreds of interesting data sets
- [Awesome datasets](https://github.com/caesar0301/awesome-public-datasets) - A list of data sets hosted on GitHub
- [rs.io](http://rs.io/100-interesting-data-sets-for-statistics/) - A great blog post with hundreds of interesting data sets
In real-world data science, you may not find an ideal data set. You might have to aggregate disparate data sources instead, or do a good amount of data cleaning.

For the purposes of this project, we'll be using data about New York City public schools, which can be found [here](https://data.cityofnewyork.us/browse?category=Education).

## 1.2 Finding All of the Relevant Data Sets

Once you've chosen a topic, you'll want to pick an angle to investigate. It's important to choose an angle that has enough depth to analyze, but isn't so complicated that it's difficult to get started. You want to finish the project, and you want your results to be interesting to others.

One of the most controversial issues in the U.S. educational system is the efficacy of standardized tests, and whether they're unfair to certain groups. Given our prior knowledge of this topic, investigating the correlations between [SAT scores](https://en.wikipedia.org/wiki/SAT) and demographics might be an interesting angle to take. We could correlate SAT scores with factors like race, gender, income, and more.

The SAT, or Scholastic Aptitude Test, is an exam that U.S. high school students take before applying to college. Colleges take the test scores into account when deciding who to admit, so it's fairly important to perform well on it.

The test consists of three sections, each of which has 800 possible points. The combined score is out of 2,400 possible points (while this number has changed a few times, the data set for our project is based on 2,400 total points). **Organizations often rank high schools by their average SAT scores.** The scores are also considered a measure of overall school district quality.

New York City makes its data on [high school SAT scores](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4) available online, as well as the [demographics for each high school](https://data.cityofnewyork.us/Education/DOE-High-School-Directory-2014-2015/n3p6-zve2). The first few rows of the SAT data look like this:

<left><img width="600" src="https://drive.google.com/uc?export=view&id=1_1f_s0jbjSH1YeWs-W3Ioy8mqDqaYYNb"></left>

Unfortunately, combining both of the data sets won't give us all of the demographic information we want to use. We'll need to supplement our data with other sources to do our full analysis.

The same website has several related data sets covering demographic information and test scores. Here are the links to all of the data sets we'll be using:

- [SAT scores by school](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4) - SAT scores for each high school in New York City
- [School attendance](https://data.cityofnewyork.us/Education/School-Attendance-and-Enrollment-Statistics-by-Dis/7z8d-msnt) - Attendance information for each school in New York City
- [Class size](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3) - Information on class size for each school
- [AP test results](https://data.cityofnewyork.us/Education/AP-College-Board-2010-School-Level-Results/itfs-ms3e) - Advanced Placement (AP) exam results for each high school (passing an optional AP exam in a particular subject can earn a student college credit in that subject)
- [Graduation outcomes](https://data.cityofnewyork.us/Education/Graduation-Outcomes-Classes-Of-2005-2010-School-Le/vh2h-md7a) - The percentage of students who graduated, and other outcome information
- [Demographics](https://data.cityofnewyork.us/Education/School-Demographics-and-Accountability-Snapshot-20/ihfw-zy9j) - Demographic information for each school
- [School survey](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) - Surveys of parents, teachers, and students at each school
All of these data sets are interrelated. We'll need to combine them into a single data set before we can find correlations.

## 1.3 Finding Background Information

Before we move into coding, we'll need to do some background research. A thorough understanding of the data will help us avoid costly mistakes, such as thinking that a column represents something other than what it does. Background research will also give us a better understanding of how to combine and analyze the data.

In this case, we'll want to research:

- [New York City](https://en.wikipedia.org/wiki/New_York_City)
- [The SAT](https://en.wikipedia.org/wiki/SAT)
- [Schools in New York City](https://en.wikipedia.org/wiki/List_of_high_schools_in_New_York_City)
- [Our data](https://data.cityofnewyork.us/browse?category=Education)

We can learn a few different things from these resources. For example:

- Only high school students take the SAT, so we'll want to focus on high schools.
- New York City is made up of five boroughs, which are essentially distinct regions.
- New York City schools fall within several different school districts, each of which can contains dozens of schools.
- Our data sets include several different types of schools. We'll need to clean them so that we can focus on high schools only.
- Each school in New York City has a unique code called a **DBN**, or district borough number.
- Aggregating data by district will allow us to use the district mapping data to plot district-by-district differences.

## 1.4 Reading in the Data

Once we've done our background research, we're ready to read in the data. For your convenience, we've placed all the data into the schools folder. Here are all of the files in the folder:

- **ap_2010.csv** - [Data on AP test results](https://data.cityofnewyork.us/Education/AP-College-Board-2010-School-Level-Results/itfs-ms3e)
- **class_size.csv** - Data on [class size](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3)
- **demographics.csv** - Data on [demographics](https://data.cityofnewyork.us/Education/School-Demographics-and-Accountability-Snapshot-20/ihfw-zy9j)
- **graduation.csv** - Data on [graduation outcomes](https://data.cityofnewyork.us/Education/Graduation-Outcomes-Classes-Of-2005-2010-School-Le/vh2h-md7a)
- **hs_directory.csv** - [A directory of high schools](https://data.cityofnewyork.us/Education/DOE-High-School-Directory-2014-2015/n3p6-zve2)
- **sat_results.csv** - Data on [SAT scores](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4)
- **survey_all.txt** - Data on [surveys](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) from all schools
- **survey_d75.txt** - Data on surveys from New York City district 75

**survey_all.txt** and **survey_d75.txt** are in more complicated formats than the other files. For now, we'll focus on reading in the - CSV files only, and then explore them.

We'll read each file into a [pandas dataframe](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html), and then store all of the dataframes in a dictionary. This will give us a convenient way to store them, and a quick way to reference them later on.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Read each of the files in the list **data_files** into a pandas dataframe using the **pandas.read_csv()** function.
- Recall that all of the data sets are in the **datasets** folder. That means the path to **ap_2010.csv** is **datasets/ap_2010.csv**.
- Add each of the dataframes to the **dictionary data**, using the base of **filename as the key**. For example, you'd enter **ap_2010** for the file **ap_2010.csv.**
- Afterwards, **data** should have the following keys:
    - **ap_2010**
    - **class_size**
    - **demographics**
    - **graduation**
    - **hs_directory**
    - **sat_results**
- In addition, each key in **data** should have the corresponding dataframe as its value.


In [0]:
# put your code here

import pandas as pd
data_files = [
    "ap_2010.csv",
    "class_size.csv",
    "demographics.csv",
    "graduation.csv",
    "hs_directory.csv",
    "sat_results.csv"
]
data = {}
for f in data_files:
  d = pd.read_csv("{0}".format(f))
  key_name = f.replace(".csv", "")
  data[key_name] = d
  print(d.shape)


(258, 5)
(27611, 16)
(10075, 38)
(25096, 23)
(435, 64)
(478, 6)


## 1.5 Exploring the SAT Data

What we're mainly interested in is the SAT data set, which corresponds to the dictionary key **sat_results**. This data set contains the **SAT scores** for each high school in New York City. We eventually want to correlate selected information from this data set with information in the other data sets.

Let's explore **sat_results** to see what we can discover. Exploring the dataframe will help us understand the structure of the data, and make it easier for us to analyze it.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Display the first five rows of the SAT scores data.
    - Use the key **sat_results** to access the SAT scores dataframe stored in the dictionary **data.**
    - Use the **pandas.DataFrame.head()** method along with the **print()** function to display the first five rows of the dataframe.

In [0]:
# put your code here

print( data["sat_results"].head() )

      DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    38

## 1.6 Exploring the Remaining Data

When we printed the first five rows of the SAT data, the output looked like this:

```python
DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL
2  01M450                     EAST SIDE COMMUNITY SCHOOL
3  01M458                      FORSYTH SATELLITE ACADEMY
4  01M509                        MARTA VALLE HIGH SCHOOL

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404
1                     91                             383                 423
2                     70                             377                 402
3                      7                             414                 401
4                     44                             390                 433

  SAT Writing Avg. Score
0                    363
1                    366
2                    370
3                    359
4                    384
```

We can make a few observations based on this output:

- The **DBN** appears to be a unique ID for each school.
- We can tell from the first few rows of names that we only have data about high schools.
- There's only a single row for each high school, so each **DBN** is unique in the SAT data.
- We may eventually want to combine the three columns that contain SAT scores -- **SAT Critical Reading Avg.**, **Score SAT Math Avg. Score**, and **SAT Writing Avg. Score** -- into a single column to make the scores easier to analyze.
Given these observations, let's explore the other data sets to see if we can gain any insight into how to combine them.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Loop through each **key** in **data**. For each key:
    - Display the first five rows of the dataframe associated with the **key**.

In [0]:
# put your code here
for key in data.keys():
  print( data[key].head() )

      DBN                             SchoolName AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.              39   
1  01M450                 EAST SIDE COMMUNITY HS              19   
2  01M515                    LOWER EASTSIDE PREP              24   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH             255   
4  02M296  High School of Hospitality Management               s   

  Total Exams Taken Number of Exams with scores 3 4 or 5  
0                49                                   10  
1                21                                    s  
2                26                                   24  
3               377                                  191  
4                 s                                    s  
   CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1

## 1.7 Reading in the Survey Data


We can make some observations based on the first few rows of each one.

- Each data set appears to either have a **DBN** column, or the information we need to create one. That means we can use a **DBN** column to combine the data sets. First we'll pinpoint matching rows from different data sets by looking for identical **DBNs**, then group all of their columns together in a single data set.
- Some fields look interesting for mapping -- particularly **Location 1**, which contains coordinates inside a larger string.
- Some of the data sets appear to contain multiple rows for each school (because the rows have duplicate **DBN** values). That means we’ll have to do some preprocessing to ensure that each **DBN** is unique within each data set. If we don't do this, we'll run into problems when we combine the data sets, because we might be merging two rows in one data set with one row in another data set.

Before we proceed with the merge, we should make sure we have all of the data we want to unify. We mentioned the survey data earlier (**survey_all.txt** and **survey_d75.txt**), but we didn't read those files in because they're in a slightly more complex format.

Each survey text file looks like this:

```python
dbn bn  schoolname  d75 studentssurveyed    highschool  schooltype  rr_s
"01M015"    "M015"  "P.S. 015 Roberto Clemente" 0   "No"    0   "Elementary School"     88
```

The files are tab delimited and encoded with **Windows-1252** encoding. An encoding defines how a computer stores the contents of a file in binary. The most common encodings are **UTF-8** and **ASCII**. **Windows-1252** is rarely used, and can cause errors if we read such a file in without specifying the encoding. If you'd like to read more about encodings, [here's](http://kunststube.net/encoding/) a good primer.

We'll need to specify the encoding and delimiter to the pandas [pandas.read_csv()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) function to ensure it reads the surveys in properly.

After we read in the survey data, we'll want to combine it into a single dataframe. We can do this by calling the [pandas.concat()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) function:

```python
z = pd.concat([x,y], axis=0)
```

The code above will combine dataframes x and y by essentially appending y to the end of x. The combined dataframe z will have the number of rows in x plus the number of rows in y.

## 1.8 Reading in the Survey Data

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Read in **survey_all.txt.**
    - Use the **pandas.read_csv()** function to read **survey_all.txt** into the variable **all_survey**. Recall that this file is located in the **datasets** folder.
    - Specify the keyword argument **delimiter="\t".**
    - Specify the keyword argument **encoding="windows-1252".**
- Read in **survey_d75.txt.**
    - Use the **pandas.read_csv()** function to read **datasets/survey_d75.txt** into the variable **d75_survey**. Recall that this file is located in the **datasets** folder.
    - Specify the keyword argument **delimiter="\t".**
    - Specify the keyword argument **encoding="windows-1252".**
- Combine **d75_survey** and **all_survey** into a single dataframe.
    - Use the pandas [concat()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) function with the keyword argument **axis=0** to combine **d75_survey** and **all_survey** into the dataframe **survey**.
    - Pass in **all_survey** first, then **d75_survey** when calling the **pandas.concat()** function.
- Display the first five rows of **survey** using the **pandas.DataFrame.head()** function.



In [0]:
# put your code here
all_survey = pd.read_csv( 'survey_all.txt', delimiter='\t', encoding='windows-1252' )
d75_survey = pd.read_csv( 'survey_d75.txt', delimiter='\t', encoding='windows-1252' )

survey = pd.concat( [all_survey, d75_survey], axis=0 )
survey.head()

N_p    N_s   N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  \
0   90.0    NaN  22.0       7.8       NaN       7.9         7.9  M015   
1  161.0    NaN  34.0       7.8       NaN       9.1         8.4  M019   
2  367.0    NaN  42.0       8.6       NaN       7.5         8.0  M020   
3  151.0  145.0  29.0       8.5       7.4       7.8         7.9  M034   
4   90.0    NaN  23.0       7.9       NaN       8.1         8.0  M063   

   com_p_11  com_s_11   ...    t_q8c_1  t_q8c_2  t_q8c_3 t_q8c_4  t_q9  \
0       7.6       NaN   ...       29.0     67.0      5.0     0.0   NaN   
1       7.6       NaN   ...       74.0     21.0      6.0     0.0   NaN   
2       8.3       NaN   ...       33.0     35.0     20.0    13.0   NaN   
3       8.2       5.9   ...       21.0     45.0     28.0     7.0   NaN   
4       7.9       NaN   ...       59.0     36.0      5.0     0.0   NaN   

   t_q9_1  t_q9_2  t_q9_3  t_q9_4  t_q9_5  
0     5.0    14.0    52.0    24.0     5.0  
1     3.0     6.0     3.0    78.0     9.0  
2     3.0     5.0    16.0    70.0     5.0  
3     0.0    18.0    32.0    39.0    11.0  
4    10.0     5.0    10.0    60.0    15.0  

[5 rows x 2773 columns]

## 1.9 Cleaning Up the Surveys

In the last step, the expected output was:

```python
    N_p  N_s  N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  com_p_11  \
0   90  NaN   22       7.8       NaN       7.9         7.9  M015       7.6   
1  161  NaN   34       7.8       NaN       9.1         8.4  M019       7.6
```

There are two immediate facts that we can see in the data:

- There are over **2000** columns, nearly all of which we don't need. We'll have to filter the data to remove the unnecessary ones. Working with fewer columns will make it easier to print the dataframe out and find correlations within it.
- The survey data has a **dbn** column that we'll want to convert to uppercase (**DBN**). The conversion will make the column name consistent with the other data sets.

First, we'll need filter the columns to remove the ones we don't need. Luckily, there's a data dictionary at the [original data download](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) location. The dictionary tells us what each column represents. Based on our knowledge of the problem and the analysis we're trying to do, we can use the data dictionary to determine which columns to use.

Here's a preview of the data dictionary:


<left><img width="800" src="https://drive.google.com/uc?export=view&id=145GxwJiDOodRgyXfV15ybUBqfbDLFwEU"></left>


Based on the dictionary, it looks like these are the relevant columns:

```python
["dbn", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", 
 "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", 
 "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11",
 "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]
```

These columns will give us aggregate survey data about how parents, teachers, and students feel about school safety, academic performance, and more. It will also give us the **DBN**, which allows us to uniquely identify the school.

Before we filter columns out, we'll want to copy the data from the **dbn** column into a new column called **DBN**. We can copy columns like this:

```python
survey["new_column"] = survey["old_column"]
```

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Copy the data from the **dbn** column of **survey** into a new column in **survey** called **DBN**.
- Filter **survey** so it only contains the columns we listed above. You can do this using **pandas.DataFrame.loc[]**.
    - Remember that we renamed **dbn** to **DBN**; be sure to change the list of columns we want to keep accordingly.
- Assign the dataframe **survey** to the key **survey** in the dictionary **data**.
- When you're finished, the value in **data["survey"]** should be a dataframe with 23 columns and 1702 rows.


In [0]:
# put your code here
# changing the dbn column name
survey['DBN'] = survey['dbn']

# filtering
survey_cp = survey.loc[ : ,['DBN',"rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", 
 "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", 
 "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11",
 "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]]

# adding the 'survey' key with the dataframe as value
data['survey'] = survey_cp

data['survey']

DBN   rr_s  rr_t  rr_p    N_s    N_t    N_p  saf_p_11  com_p_11  \
0   01M015    NaN    88    60    NaN   22.0   90.0       8.5       7.6   
1   01M019    NaN   100    60    NaN   34.0  161.0       8.4       7.6   
2   01M020    NaN    88    73    NaN   42.0  367.0       8.9       8.3   
3   01M034   89.0    73    50  145.0   29.0  151.0       8.8       8.2   
4   01M063    NaN   100    60    NaN   23.0   90.0       8.7       7.9   
5   01M064    NaN    94    61    NaN   32.0  159.0       8.8       8.2   
6   01M110    NaN    91    79    NaN   32.0  302.0       8.4       7.6   
7   01M134    NaN    74    66    NaN   23.0  192.0       9.0       7.9   
8   01M137    NaN    67    75    NaN   14.0  145.0       8.7       8.0   
9   01M140   98.0    94    68  185.0   34.0  213.0       8.8       8.0   
10  01M142    NaN   100   100    NaN   38.0  331.0       9.1       8.0   
11  01M184   96.0    82    80  156.0   36.0  382.0       8.7       8.1   
12  01M188   82.0    59    40  119.0   22.0  106.0       8.7       8.1   
13  01M292   89.0    70    39  379.0   26.0  151.0       7.8       7.7   
14  01M301   99.0   100    21  181.0   22.0   38.0       8.5       7.3   
15  01M315    NaN   100    67    NaN   20.0  135.0       9.0       8.6   
16  01M332   96.0   100    86  110.0   14.0   93.0       8.4       7.9   
17  01M345  100.0    87    53  254.0   20.0  127.0       8.4       7.6   
18  01M361    NaN    74    42    NaN   17.0   93.0       8.9       8.2   
19  01M363    NaN    54    40    NaN   13.0  115.0       8.7       8.1   
20  01M364    NaN    86    54    NaN   19.0  152.0       9.3       8.9   
21  01M378   78.0    91    82  204.0   20.0  203.0       8.8       7.9   
22  01M448   84.0    95    10  385.0   37.0   46.0       7.9       7.4   
23  01M450    0.0    98    28    NaN   42.0  150.0       8.7       8.2   
24  01M458   40.0   100    23   66.0   10.0   37.0       8.1       7.0   
25  01M509   90.0   100    21  306.0   29.0   69.0       7.7       7.4   
26  01M515   95.0   100    86  454.0   36.0  389.0       8.3       7.2   
27  01M539   98.0    68    51  923.0   67.0  736.0       8.5       7.9   
28  01M650   75.0    86    19  125.0   12.0   30.0       9.0       8.4   
29  01M696   98.0    81    50  561.0   30.0  283.0       8.8       8.2   
..     ...    ...   ...   ...    ...    ...    ...       ...       ...   
26  75M811   62.0    61    54   44.0   54.0  172.0       8.8       8.4   
27  75Q004   89.0    92    54   24.0   95.0  198.0       9.0       8.4   
28  75Q009   57.0    61    39   51.0   59.0  163.0       8.9       8.2   
29  75Q023   75.0    93    21   93.0   65.0   52.0       8.7       8.0   
30  75Q075   40.0    69    28   23.0   57.0   85.0       8.7       7.7   
31  75Q177   64.0    70    41   95.0   85.0  197.0       8.9       8.0   
32  75Q224   97.0    75    49   62.0   75.0  205.0       9.0       8.3   
33  75Q233   85.0    73    26  121.0   56.0  101.0       8.9       8.0   
34  75Q255   71.0    60    48   12.0   76.0  194.0       8.9       8.2   
35  75Q256   88.0    93    46  195.0  105.0  184.0       8.6       8.0   
36  75Q721   90.0    77    43  195.0   61.0  172.0       8.9       8.1   
37  75Q752   45.0    47     9  160.0   36.0   34.0       8.5       7.5   
38  75Q811   75.0    71    34  101.0   49.0  111.0       8.8       8.0   
39  75Q993   87.0    76    47   90.0   71.0  185.0       9.2       8.5   
40  75R025   57.0    75    21  238.0   70.0   97.0       8.1       7.5   
41  75R037    0.0    55    46    NaN   52.0  143.0       8.9       8.0   
42  75R373   92.0    60    34   22.0   64.0  132.0       8.8       8.1   
43  75R721   96.0    93    57   82.0   71.0  187.0       9.1       8.5   
44  75X010   84.0    91    56   76.0   86.0  247.0       9.0       8.5   
45  75X012   19.0    47     9   37.0   25.0   18.0       5.8       6.5   
46  75X017   69.0    79    37   64.0   72.0  122.0       8.8       8.1   
47  75X168   92.0    71    53   78.0   61.0  206.0       8.9       8.5   
48  75X176 

## 1.10 Inserting DBN Fields

When we explored all of the data sets, we noticed that some of them, like **class_size** and **hs_directory**, don't have a DBN column. **hs_directory** does have a **dbn** column, though, so we can just rename it.

However, **class_size** doesn't appear to have the column at all. Here are the first few rows of the data set:

```python
    CSD BOROUGH   SCHOOL CODE         SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED
```

Here are the first few rows of the **sat_results** data, which does have a **DBN** column:

```python
DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL
2  01M450                     EAST SIDE COMMUNITY SCHOOL
3  01M458                      FORSYTH SATELLITE ACADEMY
4  01M509                        MARTA VALLE HIGH SCHOOL
```

From looking at these rows, we can tell that the **DBN** in the **sat_results** data is just a combination of the **CSD** and **SCHOOL CODE** columns in the **class_size** data. The main difference is that the **DBN** is padded, so that the **CSD** portion of it always consists of two digits. That means we'll need to add a leading 0 to the **CSD** if the **CSD** is less than two digits long. Here's a diagram illustrating what we need to do:

<center><img width="150" src="https://drive.google.com/uc?export=view&id=1k1QuSctrJnMW6gfAbTs-DJbzTe23en2U"></center>

As you can see, whenever the **CSD** is less than two digits long, we need to add a leading 0. We can accomplish this using the **pandas.Series.apply()** method, along with a custom function that:

- Takes in a number.
- Converts the number to a string using the **str()** function.
- Check the length of the string using the **len()** function.
    - If the string is two digits long, returns the string.
    - If the string is one digit long, adds a 0 to the front of the string, then returns it.
        - You can use the string method [zfill()](https://docs.python.org/3/library/stdtypes.html#str.zfill) to do this.

Once we've padded the **CSD**, we can use the addition operator (+) to combine the values in the **CSD** and **SCHOOL CODE** columns. Here's an example of how we would do this:

```python
dataframe["new_column"] = dataframe["column_one"] + dataframe["column_two"]
```

And here's a diagram illustrating the basic concept:

<center><img width="300" src="https://drive.google.com/uc?export=view&id=1Io7o-45Pixlv2tX8rTyZvL6KjIq9qMPz"></center>


## 1.11 Inserting DBN Fields

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Copy the **dbn** column in **hs_directory** into a new column called **DBN**.
- Create a new column called **padded_csd** in the **class_size** data set.
    - Use the [pandas.Series.apply()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.apply.html) method along with the function **pad_csd()** to generate this column.
        - Make sure to apply the function along the **data["class_size"]["CSD"]** column.
- Use the addition operator (+) along with the **padded_csd** and **SCHOOL CODE** columns of **class_size**, then assign the result to the **DBN** column of **class_size**.
- Display the first few rows of **class_size** to double check the **DBN** column.

In [0]:
# put your code here
# copying dbn column to a new
data['hs_directory']['DBN'] = data['hs_directory']['dbn']

def pad_csd(num):
    string_representation = str(num)
    if len(string_representation) > 1: 
        return string_representation
    else:
        return string_representation.zfill(2)
      
# iterate over the CSD columns values and apply pad_csd function
data['class_size']['padded_csd'] = [ pad_csd( num ) for num in data['class_size']['CSD'] ]
# making the new dbn values
data['class_size']['DBN'] = data['class_size']['padded_csd'] + data['class_size']['SCHOOL CODE']

# showing the first 5 rows
data['class_size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  \
0                   19.0         ATS                             NaN   
1                   21.0         ATS                             NaN   
2                   17.0         ATS                             NaN   
3                   17.0         ATS                             NaN   
4                   15.0         ATS                             NaN   

  padded_csd     DBN  
0         01  01M015  
1         01  01M015  
2         01  01M015  
3         01  01M015  
4         01  01M015

## 1.12 Combining the SAT Scores


In [0]:
# put your code here
# converting the columns
data['sat_results']['SAT Math Avg. Score'] = pd.to_numeric(data['sat_results']['SAT Math Avg. Score'], errors='coerce')
data['sat_results']['SAT Critical Reading Avg. Score'] = pd.to_numeric(data['sat_results']['SAT Critical Reading Avg. Score'], errors='coerce')
data['sat_results']['SAT Writing Avg. Score'] = pd.to_numeric(data['sat_results']['SAT Writing Avg. Score'], errors='coerce')

# adding the three columns as sat_score
data['sat_results']['sat_score'] = data['sat_results']['SAT Math Avg. Score'] + data['sat_results']['SAT Critical Reading Avg. Score'] + data['sat_results']['SAT Writing Avg. Score']
  
# shows the first 5 rows
data['sat_results'].head()


DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                      7                            414.0   
4                     44                            390.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  
0                404.0                   363.0     1122.0  
1                423.0                   366.0     1172.0  
2                402.0                   370.0     1149.0  
3                401.0                   359.0     1174.0  
4                433.0                   384.0     1207.0

Now we're almost ready to combine our data sets. Before we do, let's take some time to calculate variables that will be useful in our analysis. We've already discussed one such variable -- a column that totals up the SAT scores for the different sections of the exam. This will make it much easier to correlate scores with demographic factors because we'll be working with a single number, rather than three different ones.

Before we can generate this column, we'll need to convert the **SAT Math Avg. Score**, **SAT Critical Reading Avg. Score**, and **SAT Writing Avg. Score** columns in the **sat_results** data set from the object (string) data type to a numeric data type. We can use the [pandas.to_numeric()](http://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.to_numeric.html) method for the conversion. If we don't convert the values, we won't be able to add the columns together.

It's important to pass the keyword argument **errors="coerce"** when we call **pandas.to_numeric()**, so that pandas treats any invalid strings it can't convert to numbers as missing values instead.

After we perform the conversion, we can use the addition operator (+) to add all three columns together.


**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Convert the **SAT Math Avg. Score**, **SAT Critical Reading Avg. Score**, and **SAT Writing Avg. Score** columns in the **sat_results** data set from the object (string) data type to a numeric data type.
    - Use the **pandas.to_numeric()** function on each of the columns, and assign the result back to the same column.
    - Pass in the keyword argument **errors="coerce".**
- Create a column called **sat_score** in **sat_results** that holds the combined SAT score for each student.
    - Add up **SAT Math Avg. Score**, **SAT Critical Reading Avg. Score**, and **SAT Writing Avg. Score**, and assign the total to the **sat_score** column of **sat_results.**
- Display the first few rows of the **sat_score column** of **sat_results** to verify that everything went okay.

## 1.13 Parsing Geographic Coordinates for Schools

Next, we'll want to parse the **latitude** and **longitude** coordinates for each school. This will enable us to map the schools and uncover any geographic patterns in the data. The coordinates are currently in the text field **Location 1** in the **hs_directory** data set.

Let's take a look at the first few rows:

```python
0    883 Classon Avenue\nBrooklyn, NY 11225\n(40.67...
1    1110 Boston Road\nBronx, NY 10456\n(40.8276026...
2    1501 Jerome Avenue\nBronx, NY 10452\n(40.84241...
3    411 Pearl Street\nNew York, NY 10038\n(40.7106...
4    160-20 Goethals Avenue\nJamaica, NY 11432\n(40...
```        
                                               
As you can see, this field contains a lot of information we don't need. We want to extract the coordinates, which are in parentheses at the end of the field. Here's an example:

```python
1110 Boston Road\nBronx, NY 10456\n(40.8276026690005, -73.90447525699966)
```

We want to extract the **latitude**, 40.8276026690005, and the **longitude**, -73.90447525699966. Taken together, **latitude** and **longitude** make up a pair of coordinates that allows us to pinpoint any location on Earth.

We can do the extraction with a **regular expression**. The following expression will pull out everything inside the parentheses:

```python
import re
re.findall("\(.+\)", "1110 Boston Road\nBronx, NY 10456\n(40.8276026690005, -73.90447525699966)")
```

This command will return (40.8276026690005, -73.90447525699966). We'll need to process this result further using the string methods [split()](https://docs.python.org/3/library/stdtypes.html#str.split) and [replace()](https://docs.python.org/3/library/stdtypes.html#str.replace) methods to extract each coordinate.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Write a function that:
    - Takes in a string
    - Uses the regular expression above to extract the coordinates
    - Uses string manipulation functions to pull out the latitude
    - Returns the latitude
- Use the [Series.apply()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.apply.html) method to apply the function across the **Location 1** column of **hs_directory**. Assign the result to the **lat** column of **hs_directory.**
- Display the first few rows of **hs_directory** to verify the results.

In [0]:
# put your code here
import re

def latitude( loca ):
  '''
    keeps a string, finds the between the brackets, than remove them and split the string by a ','.
      return the first element of the resultant list [latitute, longitude]
  '''
  lati = re.findall( '\(.+\)', loca )[0].replace('(', '').replace(')', '').split(',')
  return lati[0]

data['hs_directory']['lat'] = data['hs_directory']['Location 1'].apply( latitude )

data['hs_directory'].head()

dbn                                        school_name    borough  \
0  17K548                Brooklyn School for Music & Theatre   Brooklyn   
1  09X543                   High School for Violin and Dance      Bronx   
2  09X327        Comprehensive Model School Project M.S. 327      Bronx   
3  02M280     Manhattan Early College School for Advertising  Manhattan   
4  28Q680  Queens Gateway to Health Sciences Secondary Sc...     Queens   

  building_code    phone_number    fax_number  grade_span_min  grade_span_max  \
0          K440    718-230-6250  718-230-6262             9.0              12   
1          X400    718-842-0687  718-589-9849             9.0              12   
2          X240    718-294-8111  718-294-8109             6.0              12   
3          M520  718-935-3477             NaN             9.0              10   
4          Q695    718-969-3155  718-969-3552             6.0              12   

   expgrade_span_min  expgrade_span_max         ...         priority10  \
0                NaN                NaN         ...                NaN   
1                NaN                NaN         ...                NaN   
2                NaN                NaN         ...                NaN   
3                9.0               14.0         ...                NaN   
4                NaN                NaN         ...                NaN   

                                          Location 1 Community Board  \
0  883 Classon Avenue\nBrooklyn, NY 11225\n(40.67...             9.0   
1  1110 Boston Road\nBronx, NY 10456\n(40.8276026...             3.0   
2  1501 Jerome Avenue\nBronx, NY 10452\n(40.84241...             4.0   
3  411 Pearl Street\nNew York, NY 10038\n(40.7106...             1.0   
4  160 20 Goethals Avenue\nJamaica, NY 11432\n(40...             8.0   

  Council District Census Tract        BIN           BBL  \
0             35.0        213.0  3029686.0  3.011870e+09   
1             16.0        135.0  2004526.0  2.026340e+09   
2             14.0        209.0  2008336.0  2.028590e+09   
3              1.0         29.0  1001388.0  1.001130e+09   
4             24.0       1267.0  4539721.0  4.068580e+09   

                                                 NTA     DBN  \
0  Crown Heights South                           ...  17K548   
1  Morrisania-Melrose                            ...  09X543   
2  West Concourse                                ...  09X327   
3  Chinatown                                     ...  02M280   
4  Pomonok-Flushing Heights-Hillcrest            ...  28Q680   

                  lat  
0   40.67029890700047  
1    40.8276026690005  
2  40.842414068000494  
3   40.71067947100045  
4  40.718810094000446  

[5 rows x 66 columns]

## 1.14 Extracting the Longitude

On the last screen, we parsed the **latitude** from the **Location 1** column. Now we'll just need to do the same for the longitude.

Once we have both coordinates, we'll need to convert them to numeric values. We can use the **pandas.to_numeric()** function to convert them from strings to numbers.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Write a function that:
    - Takes in a string.
    - Uses the regular expression above to extract the coordinates.
    - Uses string manipulation functions to pull out the longitude.
    - Returns the longitude.
- Use the **Series.apply()** method to apply the function across the **Location 1** column of **hs_directory**. Assign the result to the **lon** column of **hs_directory.**
- Use the **to_numeric()** function to convert the **lat** and **lon** columns of **hs_directory** to numbers.
    - Specify the **errors="coerce"** keyword argument to handle missing values properly.
- Display the first few rows of hs_directory to verify the results.

In [0]:
# put your code here
def longitude( loca ):
  '''
    keeps a string, finds the between the brackets, than remove them and split the string by a ','.
      return the second element of the resultant list [latitute, longitude]
  '''
  long = re.findall( '\(.+\)', loca )[0].replace('(', '').replace(')', '').split(',')
  return long[1]

data['hs_directory']['lon'] = data['hs_directory']['Location 1'].apply( longitude )

data['hs_directory'].head()

dbn                                        school_name    borough  \
0  17K548                Brooklyn School for Music & Theatre   Brooklyn   
1  09X543                   High School for Violin and Dance      Bronx   
2  09X327        Comprehensive Model School Project M.S. 327      Bronx   
3  02M280     Manhattan Early College School for Advertising  Manhattan   
4  28Q680  Queens Gateway to Health Sciences Secondary Sc...     Queens   

  building_code    phone_number    fax_number  grade_span_min  grade_span_max  \
0          K440    718-230-6250  718-230-6262             9.0              12   
1          X400    718-842-0687  718-589-9849             9.0              12   
2          X240    718-294-8111  718-294-8109             6.0              12   
3          M520  718-935-3477             NaN             9.0              10   
4          Q695    718-969-3155  718-969-3552             6.0              12   

   expgrade_span_min  expgrade_span_max         ...           \
0                NaN                NaN         ...            
1                NaN                NaN         ...            
2                NaN                NaN         ...            
3                9.0               14.0         ...            
4                NaN                NaN         ...            

                                          Location 1 Community Board  \
0  883 Classon Avenue\nBrooklyn, NY 11225\n(40.67...             9.0   
1  1110 Boston Road\nBronx, NY 10456\n(40.8276026...             3.0   
2  1501 Jerome Avenue\nBronx, NY 10452\n(40.84241...             4.0   
3  411 Pearl Street\nNew York, NY 10038\n(40.7106...             1.0   
4  160 20 Goethals Avenue\nJamaica, NY 11432\n(40...             8.0   

  Council District Census Tract        BIN           BBL  \
0             35.0        213.0  3029686.0  3.011870e+09   
1             16.0        135.0  2004526.0  2.026340e+09   
2             14.0        209.0  2008336.0  2.028590e+09   
3              1.0         29.0  1001388.0  1.001130e+09   
4             24.0       1267.0  4539721.0  4.068580e+09   

                                                 NTA     DBN  \
0  Crown Heights South                           ...  17K548   
1  Morrisania-Melrose                            ...  09X543   
2  West Concourse                                ...  09X327   
3  Chinatown                                     ...  02M280   
4  Pomonok-Flushing Heights-Hillcrest            ...  28Q680   

                  lat                  lon  
0   40.67029890700047   -73.96164787599963  
1    40.8276026690005   -73.90447525699966  
2  40.842414068000494   -73.91616158599965  
3   40.71067947100045   -74.00080702099967  
4  40.718810094000446   -73.80650045499965  

[5 rows x 67 columns]

## 1.15 Next Steps

We're almost ready to combine our data sets! We've come a long way in this mission -- we've gone from choosing a topic for a project to acquiring the data to having clean data that we're almost ready to combine.

Along the way, we've learned how to:

- Handle files with different formats and columns
- Prepare to merge multiple files
- Use text processing to extract coordinates from a string
- Convert columns from strings to numbers

You'll always learn something new while working on a real-world data science project. Each project is unique, and there will always be quirks you don't quite know how to handle. The key is to be willing to try different approaches, and to have a general framework in your head for how to move from Step A to Step B.

In the next section,  we'll finish cleaning the data sets, then combine them so we can start our analysis.